Use libfmt for logging

bradallred opened the pull request:
Pretty much title. Just some QoL changes.

Summary:
- changed build option `NOCOLOR` -> `COLOR_LOG`, which is now defaulted to `true` everywhere.
- the generic "string like" `formatter` has been replaced with `format_as` and better restrictions
- we delegate formatting to `fmt::format`

## Checklist

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] I used the same coding style as the surrounding code <!-- yet to be formally defined, see issue #161 -->
- [ ] I have tested the proposed changes
- [ ] I extended the documentation, if necessary
- [ ] The proposed change builds also on our build bots (check after submission)

lynxlynxlynx commented on platforms/apple/GemRB.xcodeproj/project.pbxproj:
```diff
@@ -9108,7 +9108,7 @@
 			runOnlyForDeploymentPostprocessing = 0;
 			shellPath = /bin/bash;
-			shellScript = "echo \"$SRCROOT\"\ncmake \"$SRCROOT\"\n";
+			shellScript = "echo \"$SRCROOT\"\ncmake \"$SRCROOT\" -DNOCOLOR:BOOL=0\n";
```
COLOR_LOG?

lynxlynxlynx replied:
I suspect you want it inverted here as well.

lynxlynxlynx commented on CMakeLists.txt:
```diff
@@ -514,6 +514,10 @@ endif (STATIC_LINK)
 
 INCLUDE_DIRECTORIES(${CMAKE_CURRENT_BINARY_DIR} gemrb/includes gemrb/core platforms/${CMAKE_SYSTEM_NAME})
 
+if (NOT DEFINED COLOR_LOG)
```
I suggest you add NOCOLOR as well, for anyone requesting it from before. Another block to handle the transition.

bradallred replied:
Done

lynxlynxlynx replied:
Should be false though, since you inverted the meaning of the var.

bradallred replied:
yeah, I spotted that as soon as I hit enter 🤦🏻‍♂️

lynxlynxlynx commented on gemrb/includes/win32def.h:
```diff
@@ -32,7 +32,6 @@
 #define UNICODE
 #define _UNICODE
 #define NOGDI
-#define NOCOLOR
```
There are a few other NOCOLOR mentions in the tree that need to be changed.

bradallred replied:
Any others?

lynxlynxlynx commented:
Ok, two more related things. Now windows is forced to use colors, which makes WIN32_USE_STDIO redundant. But also, there's no logging if you check the appveyor test run.

bradallred commented:
I don't have Windows to test, but it should be using `COLOR_LOG` like everything else. I'll look at `WIN32_USE_STDIO`, but Windows logger was derived from `StdioLogWriter` before this.

lynxlynxlynx commented:
What do you mean? We clearly used it to disable colored output; this is the bit this PR removes.

lynxlynxlynx commented on .appveyor.yml:
```diff
@@ -83,7 +83,7 @@ before_build:
       $env:gl = "OpenGL"
     }
-    cmake -G "Visual Studio 16 2019" -DCMAKE_BUILD_TYPE=RelWithDebInfo -DOPENGL_BACKEND="$env:gl" -DINSOURCEBUILD=1 -DWIN32_USE_STDIO=1 -DCMAKE_INSTALL_PREFIX="$env:archive_source" -DCMAKE_TOOLCHAIN_FILE=C:/tools/vcpkg/scripts/buildsystems/vcpkg.cmake -DVCPKG_TARGET_TRIPLET="x64-windows" -A x64 -DPython_ADDITIONAL_VERSIONS="3.10" .
+    cmake -G "Visual Studio 16 2019" -DCMAKE_BUILD_TYPE=RelWithDebInfo -DOPENGL_BACKEND="$env:gl" -DINSOURCEBUILD=1 -DCOLOR_LOG=true -DCMAKE_INSTALL_PREFIX="$env:archive_source" -DCMAKE_TOOLCHAIN_FILE=C:/tools/vcpkg/scripts/buildsystems/vcpkg.cmake -DVCPKG_TARGET_TRIPLET="x64-windows" -A x64 -DPython_ADDITIONAL_VERSIONS="3.10" .
```
Again, should be false.

lynxlynxlynx replied:
You're right, it does. So it's surprising this disabling didn't work. Interesting! Maybe an ordering issue with the cmake files.

bradallred commented:
arg, the issue with windows is my copy pasta is `_O_RDONLY`

bradallred commented:
hmmm, that didn't seem to produce any output either

bradallred commented:
Windows console output is back, but uncolored... the color does work on master, which apparently is hardcoded to use color regardless of options.

bradallred commented:
@MarcelHB do you have any insight into why color doesn't work here, but does on master?

lynxlynxlynx commented:
Our ansi colors defined both the background and foreground colors. Maybe we need to be explicit with fmt colors as well? It has fg/bg https://hackingcpp.com/cpp/libs/fmt.html#terminal-styles

bradallred commented:
hmmmm, it does actually look like the `ghost_white` style is applied, just the styles for the inner text is missing. Does this work as expected on Linux?

MarcelHB commented:
Color works on Windows. It's just bold-inversion as well.

lynxlynxlynx commented:
Not on the bot, before:
https://ci.appveyor.com/project/lynxlynxlynx/gemrb/builds/47575235
after:
https://ci.appveyor.com/project/lynxlynxlynx/gemrb/builds/47577935

And only noticed just now — WARNING got shortened to WARN.

MarcelHB commented:
Hm, I also see a change when switching the palette of my terminal: It lacks the background but color is still there (it was black-on-white before), with some highlighting:
![terminal](https://github.com/gemrb/gemrb/assets/238558/e1b24b63-5a94-4a92-968a-dd8fb57f64b3)

lynxlynxlynx commented:
Both the linux and windows bot now show one inconsistency: only the opening tag brackets are colored. It's more obvious on the windows one if you select the text.

bradallred commented:
yes, unfortunately the `fmt::styled` stuff doesn't work as intuitively as I had hoped. It appears each styled item completely resets the style when it is done outputting, rather than keeping a stack of formats inside a context. I suppose the answer is to just make multiple format calls.

lynxlynxlynx commented:
Now it's all bold, unfortunately.

bradallred commented:
some reading to consider:
https://github.com/actions/runner/issues/241
https://no-color.org

lynxlynxlynx commented:
Hmm, now neither build bot has color output.

lynxlynxlynx commented on platforms/windows/GemRB.cpp:
```diff
@@ -20,15 +20,24 @@
+	GetConsoleMode(hConsole, &dwMode);
+	SetConsoleMode(hConsole, dwMode | ENABLE_VIRTUAL_TERMINAL_PROCESSING);
+	return createStdioLogWriter();
+}
```
Are you sure the console mode will be reverted? It may matter only when ctrl-cing amid a print, but still.
https://learn.microsoft.com/en-us/windows/console/console-modes

bradallred commented:
> Hmm, now neither build bot has color output.

the linux one cant until the changes to `builder.yml` are merged.

> Are you sure the console mode will be reverted? It may matter only when ctrl-cing amid a print, but still.

I don't know how windows works, does windows even unwind the stack when you do whatever their equivalent of `SIGINT` is?

bradallred commented:
ok, I think this is about ready, tho perhaps you should test it on linux and ensure the `--color` option works as expected.

If there are no objections to eliminating the build time color option I'm going to squash those two commits before merging.

lynxlynxlynx commented:
I'll try it tomorrow. It's odd that the builder change is not in effect. One thing that's missing is an update to the manpage.

lynxlynxlynx commented:
--color doesn't have any effect, which explains why it's not showing on the builder either. Still looking into why.

Overall the look might have regressed on linux, but I can't say, since the bg is black in my main terminal, which I don't access to now. I propose we don't force a black background for the leading tags:
![image](https://github.com/gemrb/gemrb/assets/121515/ff2c018d-889f-4563-a151-b89c1705427b)

lynxlynxlynx commented:
LoadFromArgs has LogColor instead of LogColors key. Mode 1 looks the best here, but 2 is autodetected.

bradallred commented:
the style is something easily changed; I didn't bother trying anything but basic colors and retrieving the old colors completely depends on what the terminal palette was set to, so we just need to come up with whatever RGB values we want.

bradallred commented:
ok, that did fix colors on the linux builder. nice to know that it builds with the branch version of the config.

lynxlynxlynx commented:
One of the reasons for the NOCOLOR stuff was to support users with fancy terminal emulators using (semi)transparent backgrounds. So it'd be best to just not set a bg color in the tags.

bradallred commented:
I've removed backgrounds and squashed the `NO_COLOR` bits. We probably do want some better styles for the truecolor capable terminals, but we can adjust those later IMO

lynxlynxlynx commented:
Looking good!
